### ISSUE TITLE
[AI][Performance] boxing route never reaches the vendor fused SDPA flash kernel (composite device-type leaf routing); math-decomposition overhead closes    after a composite override

### ISSUE THREAD

**Issue by ethanZZZZZZZZZZZZ** [flagos-ai/Torch-FL#269]
Issue Type

  - [x] Performance Issue

  AI Agent Information

  - Agent: Claude Code CLI
  - Model: Claude Opus 5
  - Session Context: Perf-tuning the torch_fl boxing route toward the vendor baseline on a fixed workload (Qwen3-0.6B bf16, batch=1, greedy decode, sdpa
    attention, single Hygon DCU). Attribution showed SDPA running its math decomposition instead of fused flash on the boxing route; a composite-level override
    restores the fused path and is delivered as a perf: commit (f1f075d) with the full report in docs/bench_qwen3_dcu_route_perf.md.

  Summary

  On the boxing route, aten::scaled_dot_product_attention never reaches the vendor fused flash kernel — it always falls into the math decomposition (+784 aten
  ops/step on a Qwen3 decode step vs the vendor's 2586), which measured ~92% of the boxing route's +6.66 ms/step gap. Per-op conf routing of the attention leaf
  ops (_scaled_dot_product_flash_attention = cuda) cannot fix it, because scaled_dot_product_attention is a composite op that routes the chosen backend's leaf by
  the query's device type before dispatch ever sees the tensors. The fix registers a PrivateUse1 implementation of the composite itself that boxes q/k/v(+mask) to
  CUDA for one fused flash call and unboxes the output. Boxing e2e goes from 0.888x vendor to 0.952x vendor tps (interleaved, drift-cancelled), meeting the >= 
  0.95x vendor acceptance line.

  Environment (for bug reports)

  <details>
  <summary>Click to expand environment details</summary>

  - Platform: DCU (Hygon), CUDA-boxing build; also MetaX/maca (same CUDA-boxing path)
  - Python: 3.10.20
  - PyTorch: 2.10.0+cpu frontend + DTK fork libtorch relinked by torch_fl (flagos device)
  - torch_fl: commit 4a95268 + the P0 override (measured state); fix commit f1f075d on branch perf/qwen3-fused-sdpa-boxing
  - Hardware: single Hygon DCU card (HIP_VISIBLE_DEVICES=0), DTK 6.3.26113
  - Model: Qwen3-0.6B bf16, attn_implementation="sdpa", eval

  Runtime config:
  HIP_VISIBLE_DEVICES=0 FA_SO_PATH=/usr/local/lib/python3.10/dist-packages/flash_attn_2_cuda.cpython-310-x86_64-linux-gnu.so \
    /opt/fl-envs/boxing/bin/python benchmarks/bench_qwen3_routes.py --device flagos
  </details>

  Reproduction

  On a Qwen3-0.6B greedy-decode step (batch=1), a TorchDispatchMode op census of the boxing route:

  import torch
  import torch_fl  # boxing route, flagos device

  Observed signature of the SDPA math decomposition on the boxing route:
  _safe_softmax x28, extra bmm x57, mul.Scalar x56, clone x56, and no
  _scaled_dot_product_flash_attention calls. Full tables in
  docs/bench_qwen3_dcu_route_perf.md.

  Expected vs Actual Behavior

  Expected: the boxing route should reach the same fused flash kernel the
  vendor route uses, with a byte-identical op census.

  Actual: the composite picks the CPU flash leaf
  (_scaled_dot_product_flash_attention_for_cpu) because the tensors are
  flagos (PrivateUse1), so SDPA always falls into the math decomposition on the
  boxing route. Op-level boxing sdpa_gqa_decode measured 275.9 us vs vendor
  57.9 us (4.8x). E2E boxing was 20.94 tps vs vendor 23.57 same-session =
  0.888x.

  Root Cause Analysis
  
  aten::scaled_dot_product_attention is a composite op
  (ATen/native/transformers/attention.cpp). Its fused-backend selection runs
  inside the composite, then each chosen backend's leaf is dispatched by
  query_.device().type(): only CUDA/XPU tensors reach
  at::_scaled_dot_product_flash_attention; every other device (including
  PrivateUse1) is sent to the CPU kernel. So on the boxing route the composite
  selects the CPU flash leaf before dispatch ever sees the flagos tensors —
  no per-op routing of the leaf ops, and no _fused_sdp_choice DispatchStub
  PrivateUse1 slot (the stub only picks a backend; the leaf is still routed by
  device type), can reach the vendor fused kernel.

  Proposed Solution
  
  Register a PrivateUse1 kernel for the composite op itself
  (csrc/aten/sdp_choice_stub.cc): one DeviceBoxingGuard round-trip boxes
  q/k/v(+mask) to CUDA, the native composite runs on CUDA tensors — the vendor's
  own backend selector then picks flash exactly as on the vendor route — and the
  output is unboxed. Numerics match vendor by construction; shapes the selector
  refuses fall back to the math path as on CUDA. Guarded to CUDA-boxing builds
  (excluded on Ascend/MUSA/GCU/BPU). Grad-mode inputs fall through to the
  pre-change path (inference scope).

  Verification Plan

  - [x] Numerics vs CPU fp64 ref: all pass (bf16 max|diff| 1.998e-03…8.445e-03, tol 0.05; fp32 8.649e-07); fp32 last-token logits bitwise identical to vendor
    (max|diff|=0.0)
  - [x] Op dispatch suite: pytest tests/integration/ops -q -m anyplatform → 590 passed, 7 skipped, 4 xpassed, 2 failed (both failures pre-existing at HEAD in
    test_flaggems_conf_consistency, unrelated)
  - [x] Training smoke: backward no exception, grads defined+finite (nonzero-grad on boxing is a pre-existing cpu_fallback limitation)
  - [x] Op-level: boxing sdpa_gqa_decode 275.9 → 60.9 us (1.10x vendor)
  - [x] Op census: boxing decode 3370 → 2586 ops/step, byte-identical to vendor
  - [x] E2E (interleaved 4x4, drift-cancelled): boxing e2e 0.952x vendor (range 0.939–0.970), fwd 47.7 → 43.8 ms/step

  Context & Investigation

  1. Re-baselined all three routes at the current commit; recorded the gap table.
  2. Profiled/census'd a vendor decode step vs boxing — isolated the +784-op/step SDPA math-decomposition signature as the driver.
  3. Attempted the _fused_sdp_choice DispatchStub PrivateUse1 slot and per-op conf routing; both proven ineffective for the reason above (documented in the file
     header).
  4. Implemented the composite override, ran the correctness gates, then op-level and e2e measurement on a clean card (a stuck 29-hour orphan profiler sharing the
     card was found and killed first).

  Related Code Locations

  - csrc/aten/sdp_choice_stub.cc (new) — the composite override + rationale
  - docs/bench_qwen3_dcu_route_perf.md (new) — baseline, attribution, lever, measurement protocol
  - benchmarks/bench_qwen3_routes.py (new) — three-route e2e bench/reproducer driver
  - csrc/aten/register.cc — the WrapperMatmul pattern the override reuses

  Checklist - AI Agents MUST Complete All

  - [x] I have provided complete environment information
  - [x] I have included a minimal, self-contained reproducer
  - [x] I have included full error output with traceback
  - [x] I have analyzed the root cause (not just symptoms)
  - [x] I have proposed a specific solution with implementation approach
  - [x] I have identified affected code locations with line numbers
  - [x] I have described how to verify the fix
  - [x] I have checked for duplicate issues
  - [x] All text is in English (required per CLAUDE.md)
  - [x] Code follows project conventions (checked existing code style)

**Comment by lvyufeng** [comment]
Closing: the fix for this issue is merged.

**Fix PR:** #270 — `perf: restore fused SDPA on the boxing route via a composite override` (merge commit `400cf865`, merged 2026-09-10), which added `csrc/aten/sdp_choice_stub.cc`, `docs/bench_qwen3_dcu_route_perf.md` and `benchmarks/bench_qwen3_routes.py`. The PR carried no closing reference for this issue, so the issue stayed open with no comments. Upstream `main` is now `d210cf6`; the same file gained the FlagGems joint-attention route with its mask clause (#353) and the `AutogradPrivateUse1` registration for #383 (#390) on top of it, neither of which moves the default boxing route this issue is about.

The acceptance this close rests on is the issue's own, measured on a single Hygon DCU: boxing e2e 0.888x -> 0.952x vendor tps (interleaved, drift-cancelled), op-level `sdpa_gqa_decode` 275.9 -> 60.9 us (1.10x vendor), decode op census 3370 -> 2586 ops/step, numerics vs CPU fp64 inside tolerance, fp32 last-token logits bitwise identical to vendor.

**MetaX re-measurement — the environment note "also MetaX/maca (same CUDA-boxing path)" does not hold.** Measured on the C550 (torch `2.10.0+metax3.8.1.0`, upstream `main` @ `d210cf6`, `TORCH_ALLOW_TF32_CUBLAS_OVERRIDE` unset, which this image sets by default — see #253):

- bf16 `(1, 8, 512, 128)` on the boxing route takes the math decomposition: 21 ops (`_safe_softmax`, 2x `bmm`, `mul.Scalar`, plus the boxing `_to_copy`/`view`/`expand`), 282.2 us/call against 53.8 us/call for the same shape and the same inputs on CUDA tensors in the same process through `F.scaled_dot_product_attention` (5.24x; `max|diff| 1.953e-03`, rel `4.33e-03`, i.e. bf16-level and not a numerics bug).
- Cause, measured, not inferred: the same call on plain CUDA tensors through `torch.ops.aten.scaled_dot_product_attention.default` also produces those 21 ops, so the composite itself has no fused backend in this wheel (`UserWarning: Torch was not compiled with flash attention`, `sdp_utils.cpp:815`; no memory-efficient kernel either, `sdp_utils.cpp:866`). The fused path the CUDA route uses comes from MACA's Python replacement of `F.scaled_dot_product_attention` (`torch/nn/functional.py:6094-6135`, marked `# USE_MACA`), which calls `flash_attn.flash_attn_func` from the `flash_attn_2_cuda` extension when `query.is_cuda`. Setting `torch.disable_flash_attn = True` makes the CUDA route emit byte-for-byte the boxing route's 21-op math census, which is what pins the mechanism.
- Consequence for this issue: on MetaX the composite override cannot deliver the expected behavior by construction — the fused kernel is not inside ATen there, unlike DCU, where DTK's adapter is (`FLAGOS_DCU_SDPA_FLASH` / `torch._C._set_rocm_fa_preferred_backend`). Shapes inside `FlagGemsEligible()` are unaffected and still reach the FlagGems Triton route (spy on `flag_gems.scaled_dot_product_attention` hit once, sum `-107.6566`).

Closing as completed for the environment this issue was measured in (single Hygon DCU). The MetaX gap is filed separately as #394 rather than left implicit in this issue's environment line.
